### ISSUE TITLE
Rename the Rename Functions in std.fs to Move

### ISSUE THREAD

**Issue by mdkdir** [ziglang/zig#8719]
This is a wishlist issue, but one I hope to hear some feedback on and votes on if *you* would like to see it implemented (not if you think it's a good idea or not), regardless of what the powers that be decree.

The rename functions found in std.fs are more accurately and intuitively called move. For example:
std.fs.Dir.rename → std.fs.Dir.move
std.fs.renameAbsoluteZ → std.fs.moveAbsoluteZ

My reasoning (and my intuition) being that a file's path is unrelated to it's name, such as in foo/bar.txt, the name of the file is bar.txt, and there should be no way my re*naming*  of it should effect the path in any way. This can be a cause of confusion for those just learning about the rename function, trying to rename the file foo/bar.txt to bar.json, the function name might fool them into thinking just the name will change, not that it'll get *moved* to the current directory.

Resolving this issue also allows transfer of knowledge from the environment.
In Windows and I hope most Linux distros use the `rename` shell program to change file's name without modifying the file's position and the `move` and `mv` programs to move the file, effectively changing it's path (which is a combination of the file's location and name properties).

This isn't to say other programming languages, environments, and other programs don't use rename when they modify the file path (such as C and C++).

As I've said in the beginning, please +1 or -1 this issue using the reactions, and if you have any arguments for, against, or both, please contribute to the conversation as it does provide valuable insight into the Zig community as well as the direction of the language.

**Comment by ehaas** [comment]
`std.fs.Dir.rename` uses the `rename` family of syscall/libC functions on Linux, which require the two paths to be on the same mounted filesystem. I think naming it `move` would be equally misleading since you wouldn't be able to use it to move a file across filesystems.

**Comment by jedisct1** [comment]
Nod. `move` suggests two things:

- That the target location is in a different directory
- And that it will work across different mount points, copying the file if necessary.

The first property doesn't need to be satisfied, and the second one would be very useful, and a `move` function would be a great place to do it.

The `rename` function calls the `rename(2)` (actually `renameat(2)`) system call, which has different properties:

- It doesn't work across mount points.
- On local filesystems, it is atomic.

People looking for these properties on a Unix system are likely going to look for something called `rename`. 

So, maybe we shouldn't rename (!) that function, but either add a `move` counterpart, or add an extra parameter to get the `move` semantics if needed.

And improving the current function documentation to clearly document its properties can also be useful.

**Comment by kubkon** [comment]
I agree with @jedisct1 that `rename` is actually a very intuitive name on *nix platforms that is taken pretty much verbatim from POSIX, and as such, I am strongly against renaming it to anything different. Note that Rust's libstd also favours `std::fs::rename` over `move`, etc.

As far as adding a function that works across mount points, I'm trying to understand if this is really needed? Note that I believe we already have half of the required functionality in the form of [`std.fs.Dir.copyFile`](https://github.com/ziglang/zig/blob/2d119677344c4ca6a2700b8de0a24215b9a37c06/lib/std/fs.zig#L1943) function which will copy the file across mount points. In fact, this is the function we use to do an in-place move of a MachO binary to a new inode after it's been cached by the kernel and otherwise would lead to a sigkill.

**Comment by jedisct1** [comment]
`move` could try a `rename()` call, and only if that fails with `EXDEV`, do a copy+unlink sequence.

On Stack Overflow, there are quite a few occurrences of people confused about how to move files between different Windows drives in various languages. So, maybe there's a place for such a function.

In order to work around Apple bugs, a forced copy would still be necessary, though :)
